src/metabase/upload.clj

```diff
@@ -533,59 +539,75 @@
 
   Throws an exception if:
     - the CSV file contains duplicate column names
     - the schema of the CSV file does not match the schema of the table"
   [table header]
   ;; Assumes table-cols are unique when normalized
-  (let [fields                   (t2/select :model/Field :table_id (:id table))
-        fields-by-normed-name    (-> (into {}
-                                           (map (fn [field]
-                                                  [(normalize-column-name (:name field)) field]))
-                                           fields)
-                                     ;; ignore the _mb_row_id field, it will be filled automatically
-                                     (dissoc "_mb_row_id"))
-        normalized-field-names   (keys fields-by-normed-name)
+  (let [fields (t2/select :model/Field :table_id (:id table))
+        fields-by-normed-name (m/index-by (comp normalize-column-name :name)
+                                          fields)
+        normalized-field-names (keys fields-by-normed-name)
+        auto-pk-col-indices (set (indices-where #(= auto-pk-column-name (normalize-column-name %)) header))
+        ;; this function removes the auto-pk columns from each row and the header
+        remove-auto-pk-cols (fn [row]
+                              (remove-indices auto-pk-col-indices row))
         ;; TODO: use this to create nice error messages when there are extra or missing columns
-        normalized-header+header (into []
-                                       (map (fn [col]
-                                              [(normalize-column-name col) col]))
-                                       header)
-        normalized-header        (map first normalized-header+header)
-        ;; check for duplicates
-        _ (when (some #(< 1 %) (vals (frequencies normalized-header)))
-            (throw (ex-info (tru "The CSV file contains duplicate column names.")
-                            {:status-code 422})))
+        normalized-header+header (->> header
+                                      (map (fn [col]
+                                             [(normalize-column-name col) col]))
+                                      remove-auto-pk-cols)
+        normalized-header (map first normalized-header+header)
         [extra missing _both] (data/diff (set normalized-header)
-                                         (set normalized-field-names))]
-
+                                         ;; ignore the auto-pk field, it will be filled automatically
+                                         (disj (set normalized-field-names) auto-pk-column-name))]
+    ;; check for duplicates
+    (when (some #(< 1 %) (vals (frequencies normalized-header)))
+      (throw (ex-info (tru "The CSV file contains duplicate column names.")
+                      {:status-code 422})))
     (if (and (empty? extra) (empty? missing))
       {:parse-rows (fn [rows]
```

Comment:
This is duplicated with what we have in `detect-schema`.

it makes me think about whether we should stop having these functions returning a lexical scoped function like this.

WDYT about having a function
`(defn parse-rows [col->types rows excluded-columns-idx])`

this way we can decouple schema checking and parsing the rows.

Reply:
You have a point, ideally these shouldn't be coupled and duplicated logic can be separated. But let's think about this. `check-schema` is doing a lot here, and its logic can be split.

Schema checking and parsing the rows are coupled in the sense that: 
(1) both the header and rows need to have their auto-pk columns removed.
(2) the schema checking uses the types of fields to know what parser to apply to each column

Creating `(defn parse-rows [col->types rows excluded-columns-idx])` might help reduce some coupling because the coupling from (1) is more explicitly in the form of the `excluded-columns-idx` argument. But they're still coupled by the shared decision to remove the auto-pk columns. To decouple them fully, I could remove the auto-pk column from both the `header` argument of `schema-check` and the `rows` argument of `parse-rows`.

```
(defn auto-pk-col-indices [header])
(defn schema-check [table-fields header-without-auto-pk-columns])
(defn parse-rows [col-upload-types rows-without-auto-pk-columns])
```

Then have them called like so in `append-csv!*`:

```clojure
          excluded-columns-indices (auto-pk-col-indices header)
          header' (remove-indices excluded-columns-indices header)
          table-fields (t2/select :model/Field :table_id (:id table) :active true)
          table-has-auto-pk? (contains? fields-by-normed-name auto-pk-column-name)
          fields-by-normed-name (m/index-by (comp normalize-column-name :name) fields)
          col-upload-types (map (comp base-type->upload-type :base_type fields-by-normed-name normalize-column-name) header)
          _ (check-schema table-fields header) ;; all this does is raise errors
          ... then later
          parsed-rows (parse-rows col-upload-types rows)
```

Now (1) and (2) are both outside `check-schema`.

Reply:
awesome, I think this is much better :)